docs/tools/screenshots.sh

```diff
@@ -3,26 +3,50 @@
 # $1: python script to run
 # urxvt, xdotool and import are required to run this script
 
-CLASSNAME=$(head -c 6 /dev/urandom | base64 | tr -cd [:alnum:])
+CLASSNAME=urwid-docs-$(head -c 6 /dev/urandom | base64 | tr -cd '[:alnum:]')
 PYTHON=python
 
-urxvt -bg gray90 -b 0 +sb -fn '-misc-fixed-medium-*-*-*-*-140-*-*-*-*-*-*' \
-	-fb '-misc-fixed-bold-*-*-*-*-140-*-*-*-*-*-*' \
-	-name "$CLASSNAME" -e "$PYTHON" "$1" &
+urxvt -bg gray90 -b 0 +sb \
+	-fn 'xft:Monospace:pixelsize=20' \
+	-fb 'xft:Monospace-Bold:pixelsize=20' \
+	-name "$CLASSNAME" \
+	-e "$PYTHON" "$1" &
 RXVTPID=$!
+trap 'kill $RXVTPID' EXIT
 until RXVTWINDOWID=$(xdotool search --classname "$CLASSNAME"); do
 	sleep 0.1
 done
 export RXVTWINDOWID
 image=${1%.py}
 
+# Synchronize to a point when the example has finished rendering.
+# Once in n consecutive polls we don't see any "R (running)", consider that the
+# process has completed the initial rendering & processing xdotool events,
+# and has become idle.
+# With increasing n, the probability of misdetection falls exponentially.
+sync_example_settled() {
+	pid=$1
+	n=0
+	while (( n < 80 )); do
+		sleep 0.005
+		procstate="$(cut -d' ' -f3 < /proc/"$pid"/stat)"
+		case $procstate in
+			R) n=0; printf '!';;
+			*) _=$(( ++n )); printf '.';;
+		esac
+	done
+	echo
+}
+
+# target script is interpreted by python, use it's PID
+example_pid="$(pgrep python)"
```

Comment:
in case of `ps` usage, collected 2 PIDs with whitespaces before and `/proc/` path became invalid

Reply:
1. typo `it's` → `its`

Reply:
2. I don't understand the change; `urxvt` is not expected to have 2 child processes. See line 13. What's going on here?

Suggesting to pass `--parent $RXVTPID` to make the `pgrep` call do equivalent thing to what `ps` did

Reply:
`ps --no-headers -o pid --ppid $RXVTPID` output (from real call):
```
  47163
  47167
```
Notice also spaces before PID.

Reply:
Spaces — got it. Prolly `ps` changed since the script was written. It doesn't guarantee stable output format.

But what are those 2 processes? Can you do `ps f 47163 47167` ?

Reply:
I added to generation:
```shell
echo "Executing: ps f --ppid $RXVTPID"
ps f --ppid $RXVTPID
```
And got:
```
Executing: ps f --ppid 54950
    PID TTY      STAT   TIME COMMAND
  54956 pts/1    Rs+    0:00 python docs/examples/bigtext.py
  54952 pts/0    S+     0:00 urxvt -bg gray90 -b 0 +sb -fn xft:Monospace:pixelsize=20 -fb xft:Monospace-Bold:pixelsize=20 -name urwid-docs-lb7HHzm6 -e python docs/examples/bigtext.py
````

Reply:
@ulidtko 
probably `ps jf` will explain more:
```
   PPID     PID    PGID     SID TTY        TPGID STAT   UID   TIME COMMAND
      0       1       1       1 pts/0      17766 Ss       0   0:00 bash
      1   17766   17766       1 pts/0      17766 S+       0   0:00 make
  17766   17769   17766       1 pts/0      17766 S+       0   0:00  \_ /bin/bash -e tools/compile_pngs.sh examples/bigtext.py
  17769   17772   17766       1 pts/0      17766 S+       0   0:00      \_ Xvfb :5 -screen 0 1024x768x24
  17769   17773   17766       1 pts/0      17766 S+       0   0:00      \_ /bin/bash tools/screenshots.sh examples/bigtext.py
  17773   17778   17766       1 pts/0      17766 S+       0   0:00          \_ urxvt -bg gray90 -b 0 +sb -fn xft:Monospace:pixelsize=20 -fb xft:Monospace-Bold:pixelsize=20 -name urwid-docs-U
  17778   17780   17766       1 pts/0      17766 S+       0   0:00          |   \_ urxvt -bg gray90 -b 0 +sb -fn xft:Monospace:pixelsize=20 -fb xft:Monospace-Bold:pixelsize=20 -name urwid-do
  17778   17784   17784   17784 pts/1      17784 Ss+      0   0:00          |   \_ python examples/bigtext.py
  17773   17787   17766       1 pts/0      17766 R+       0   0:00          \_ ps jf
```

Reply:
Aha, I see, makes sense now: urxvt started doing some multiprocessing too, it has 2 pids 54950 and 54952.

Then, `pgrep --parent $RXVTPID python` as you wrote, is correct.